torchvision/prototype/models/depth/stereo/raft_stereo.py

```diff
@@ -0,0 +1,711 @@
+        corr_volume = _corr_channel_normalize(corr, num_channels)
+
+        corr_volume = corr_volume.reshape(batch_size * h * w, 1, 1, w)
+        corr_pyramid = [corr_volume]
+        for _ in range(self.num_levels - 1):
+            corr_volume = F.avg_pool2d(corr_volume, kernel_size=(1, 2), stride=(1, 2))
+            corr_pyramid.append(corr_volume)
+
+        return corr_pyramid
+
+
+class CorrBlock1d(nn.Module):
+    """The row-wise correlation block.
+
+    Use indexes from correlation pyramid to create correlation features.
+    The "indexing" of a given centroid pixel x' is done by concatenating its surrounding row neighbours
+    within radius
+    """
+
+    def __init__(self, *, num_levels: int = 4, radius: int = 4):
+        super().__init__()
+        self.radius = radius
+        self.out_channels = num_levels * (2 * radius + 1)
+
+    def forward(self, centroids_coords: Tensor, corr_pyramid: List[Tensor]) -> Tensor:
+        """Return correlation features by indexing from the pyramid."""
+        neighborhood_side_len = 2 * self.radius + 1  # see note in __init__ about out_channels
+        di = torch.linspace(-self.radius, self.radius, neighborhood_side_len).to(centroids_coords.device)
+        di = di.view(1, 1, neighborhood_side_len, 1).to(centroids_coords.device)
+
+        batch_size, _, h, w = centroids_coords.shape  # _ = 2 but we only use the first one
+        # We only consider 1d and take the first dim only
+        centroids_coords = centroids_coords[:, :1].permute(0, 2, 3, 1).reshape(batch_size * h * w, 1, 1, 1)
+
+        indexed_pyramid = []
+        for corr_volume in corr_pyramid:
+            x0 = centroids_coords + di  # end shape is (batch_size * h * w, 1, side_len, 1)
+            y0 = torch.zeros_like(x0)
+            sampling_coords = torch.cat([x0, y0], dim=-1)
+            indexed_corr_volume = grid_sample(corr_volume, sampling_coords, align_corners=True, mode="bilinear").view(
+                batch_size, h, w, -1
+            )
+            indexed_pyramid.append(indexed_corr_volume)
+            centroids_coords = centroids_coords / 2
+
+        corr_features = torch.cat(indexed_pyramid, dim=-1).permute(0, 3, 1, 2).contiguous()
+
+        expected_output_shape = (batch_size, self.out_channels, h, w)
+        torch._assert(
+            corr_features.shape == expected_output_shape,
+            f"Output shape of index pyramid is incorrect. Should be {expected_output_shape}, got {corr_features.shape}",
+        )
+        return corr_features
+
+
+class RaftStereo(nn.Module):
+    def __init__(
+        self,
+        *,
+        feature_encoder: FeatureEncoder,
+        context_encoder: MultiLevelContextEncoder,
+        corr_pyramid: CorrPyramid1d,
+        corr_block: CorrBlock1d,
+        update_block: MultiLevelUpdateBlock,
+        depth_head: nn.Module,
+        mask_predictor: Optional[nn.Module] = None,
+        slow_fast: bool = False,
+        num_iters: int = 12,
```

Comment:
This needs to be removed as discussed yesterday, in favour of moving it to forward to make it possible to change during inference and align with RAFT.